build-system/common/esbuild-babel.js

```diff
@@ -48,57 +99,44 @@ function getEsbuildBabelPlugin(
   preSetup = () => {},
   postLoad = () => {}
 ) {
-  function sha256(contents) {
+  function sha256(contents, options) {
     if (!enableCache) {
       return '';
     }
     const hash = crypto.createHash('sha256');
-    hash.update(callerName);
+    hash.update(JSON.stringify(options));
     hash.update(contents);
     return hash.digest('hex');
   }
 
   function batchedRead(path) {
     let read = readCache.get(path);
     if (!read) {
-      read = fs.promises
-        .readFile(path)
-        .then((contents) => {
-          return {
-            contents,
-            hash: sha256(contents),
-          };
-        })
-        .finally(() => {
-          readCache.delete(path);
-        });
+      read = fs.promises.readFile(path).finally(() => {
+        readCache.delete(path);
+      });
       readCache.set(path, read);
     }
     return read;
   }
 
-  function transformContents(filepath, contents, hash) {
+  async function transformContents(contents, babelOptions) {
+    const hash = sha256(contents, babelOptions);
     if (enableCache) {
-      const cached = transformCache.get(filepath);
-      if (cached && cached.hash === hash) {
-        return cached.promise;
+      const cached = await transformCache.get(hash);
```

Comment:
This `await` create a race where multiple transforms of the same file can happen. The reason this entire function was sync was to avoid it.